default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1573 @@
+            del self.design_id_by_name[corrupted_entry]
+
+    def _update_buildable_items_this_turn(self, verbose=False):
+        """Calculate which parts and hulls can be built on each planet this turn.
+
+        :param verbose: bool - toggles detailed debugging output.
+        """
+        # The AI currently has no way of checking building requirements of individual parts and hulls directly.
+        # It can only check if we can build a design. Therefore, we use specific testdesigns to check if we can
+        # build a hull or part.
+        # The building requirements are constant so calculate this only once at the beginning of each turn.
+        #
+        # Code structure:
+        #   1. Update hull test designs
+        #   2. Get a list of buildable ship hulls for each planet
+        #   3. Update ship part test designs
+        #   4. Cache the list of buildable ship parts for each planet
+        #
+        self.hulls_for_planets.clear()
+        self.parts_for_planets.clear()
+        get_shipdesign = fo.getShipDesign
+        get_hulltype = fo.getHullType
+        empire = fo.getEmpire()
+        empire_id = empire.empireID
+        universe = fo.getUniverse()
+        planets_with_shipyards = _get_planets_with_shipyard()
+        available_hulls = list(empire.availableShipHulls)   # copy so we can sort it locally
+        # Later on in the code, we need to find suitable testhulls, i.e. buildable hulls for all slottypes.
+        # To reduce the number of lookups, move the hardcoded TESTDESIGN_PREFERED_HULL to the front of the list.
+        # This hull should be buiildable on each planet and also cover the most common slottypes.
+        try:
+            idx = available_hulls.index(TESTDESIGN_PREFERED_HULL)
+            available_hulls[0], available_hulls[idx] = available_hulls[idx], available_hulls[0]
+        except ValueError:
+            print "WARNING: Tried to use '%s' as testhull but it is not in available_hulls." % TESTDESIGN_PREFERED_HULL,
+            print "Please update ShipDesignAI.py according to the new content."
+            traceback.print_exc()
+        testdesign_names = [get_shipdesign(design_id).name(False) for design_id in empire.allShipDesigns
+                            if get_shipdesign(design_id).name(False).startswith(TESTDESIGN_NAME_BASE)]
+        testdesign_names_hull = [name for name in testdesign_names if name.startswith(TESTDESIGN_NAME_HULL)]
+        testdesign_names_part = [name for name in testdesign_names if name.startswith(TESTDESIGN_NAME_PART)]
+        available_slot_types = {slottype for slotlist in [get_hulltype(hull).slots for hull in available_hulls]
+                                for slottype in slotlist}
+        new_parts = [_get_part_type(part) for part in empire.availableShipParts
+                     if part not in self.strictly_worse_parts]
+        pid = self.production_cost.keys()[0]  # TODO: Check if location invariant. If not, implement different handling
```

Comment:
- [ ] This line generates an index out of range exception if production_cost is empty, which I just observed in a game -- the AI was down to a single outpost.  With current game content there was probably not anything the AI could have done anyway, so this is not a critical error, but it is easy to fix, and this kind of issue should be kept in mind throughout.  I am also experimenting with a checkbox here to indicate when this has been taken care of :smiley: